rfcs/20180626-tensor-forest.md

```diff
@@ -0,0 +1,255 @@
+- __Disadvantages__:
+    - Need much larger ensembles due to the fact that trees are independent
+    - Can't easily handle custom losses like ranking
+    - Can have worse performance than boosted trees for a very complicated decision boundary
+
+## Algorithm
+
+[Extremely Randomized Forest](http://www.montefiore.ulg.ac.be/~ernst/uploads/news/id63/extremely-randomized-trees.pdf) is an online training algorithm, that makes quick split decisions.
+In contrast with a classic random forest, in extremely randomized forests:
+
+- Split candidates are generated after seeing only a number of samples from the data (as opposed to seeing the full dataset or a large portion, determined by the bagging fraction, of it.
+- Splits quality is evaluated over the samples of the data, as opposed to full or a large portion of the dataset.
+
+Those modifications allow to make quick decisions and forest can be grown in online fashion, and experiments demonstrate the ERF provide similar performance to that of classical forests (at the cost of having to potentially build much deeper trees).
+
+At the start of training, the tree structure is initialized to a root node, and the leaf and growing statistics for it are both empty. Then, for each batch `{(x_i, y_i)}` of training data, the following steps are performed:
+
+1. Given the current tree structure, each instance `x_i` is used to find the leaf assignment `l_i` where this instance falls into.
+2. `Y_i` (the label of `x_i`) is used to update the leaf statistics of leaf `l_i`.
+3. If the growing statistics for the leaf `l_i` do not yet contain `num_splits_to_consider` splits, `x_i` is used to generate another split. Specifically, a random feature value is chosen, and `x_i`'s value at that feature is used for the split's threshold.
+4. Otherwise, `(x_i, y_i)` is used to update the statistics of every split in the growing statistics of leaf `l_i`. If leaf `l_i` has now seen `split_after_samples` data points since creating all of its potential splits, the split with the best score is chosen, and the tree structure is grown.
+
+
+## BenchMark
+
+Comparing with Scikit-learn ExtraTrees. Both using 100 trees with 10k nodes. And Scikit-learn ExtraTrees is a batch algorithm while TensorForest is a streaming/online algorithm.
+
+|Data Set| #Examples | #Features| #Classes| TensorForest Accuracy(%)/R^2 Score| Scikit-learn ExtraTrees Accuracy(%)/R^2 Score|
+|-------|:---------:|:---------:|:---------:|:---------:|---------:|
+| Iris| 150| 4| 3| 95.6| 94.6|
+|Diabetes| 442| 10| Regression| 0.462| 0.461|
+|Boston| 506| 13| Regression| 0.793| 0.872|
+|Digits| 1797| 64| 10| 96.7| 97.6|
+|Sensit(Comb.)| 78k| 100| 3| 81.0| 83.1|
+|Aloi| 108K| 128| 1000| 89.8| 91.7|
+|rcv1| 518k| 47,236| 53| 78.7| 81.5|
+|Covertype| 581k| 54| 7| 83.0| 85.0|
+|HiGGS| 11M| 28| 2| 70.9| 71.7|
+
+With single machine training, TensorForest finishes much faster on big dataset like HIGGS, takes about one percent of the time scikit-lean required.
```

Comment:
hm where is time n this table? It is just performance metrics right?

Reply:
yeah, it's just performance metrics, i took it from the workshop paper.

Reply:
But what i am saying that this statement "much faster on big dataset" is not substantiated by this table. You either keep the table and say that it is from resource A, demonstrating that the quality is on par with scikit learn, and removing the statement that says that it trains faster. Or add a reference to the resource which states that it trains faster

Reply:
i see.. i'll add a source to the workshop paper, as the claim was also from the paper.